src/main/kotlin/org/rust/ide/highlight/RsHighlightExitPointsHandlerFactory.kt

```diff
@@ -0,0 +1,104 @@
+ * found in the LICENSE file.
+ */
+
+package org.rust.ide.highlight
+
+import com.intellij.codeInsight.highlighting.HighlightUsagesHandlerBase
+import com.intellij.codeInsight.highlighting.HighlightUsagesHandlerFactoryBase
+import com.intellij.openapi.editor.Editor
+import com.intellij.psi.PsiElement
+import com.intellij.psi.PsiFile
+import com.intellij.util.Consumer
+import org.rust.lang.core.psi.*
+import org.rust.lang.core.psi.RsElementTypes.*
+import org.rust.lang.core.psi.ext.ancestors
+import org.rust.lang.core.psi.ext.elementType
+
+class RsHighlightExitPointsHandlerFactory : HighlightUsagesHandlerFactoryBase() {
+    override fun createHighlightUsagesHandler(editor: Editor, file: PsiFile, target: PsiElement): HighlightUsagesHandlerBase<*>? {
+        if (file !is RsFile) {
+            return null
+        }
+        val elementType = target.elementType
+        if (elementType == RETURN || elementType == Q) {
+            return RsHighlightExitPointsHandler(editor, file, target)
+        }
+        return null
+    }
+
+    private class RsHighlightExitPointsHandler(editor: Editor, file: PsiFile, var target: PsiElement) : HighlightUsagesHandlerBase<PsiElement>(editor, file) {
+        override fun getTargets() = listOf(target)
+
+        override fun selectTargets(targets: MutableList<PsiElement>?, selectionConsumer: Consumer<MutableList<PsiElement>>?) {
+            selectionConsumer?.consume(targets)
+        }
+
+        private fun getFunctionOrLambda(target: PsiElement): PsiElement? =
+            target.ancestors.first { it is RsFunction || it is RsLambdaExpr }
+
+        private fun onlyExpressionAboveUntilFn(expr: RsExpr): Boolean {
+            expr.ancestors.forEach {
+                val parent = it.parent
+                when (it) {
+                    is RsExpr -> when (parent) {
+                        is RsMatchExpr,
+                        is RsPatRange,
+                        is RsMatchArmGuard -> return false
+                    }
+                    is RsFunction,
+                    is RsLambdaExpr -> return true
+                    is RsCondition,
+                    is RsStmt -> return false
+                }
+            }
+            return false
+        }
+
+        override fun computeUsages(targets: MutableList<PsiElement>?) {
+            val function: PsiElement? = getFunctionOrLambda(target)
+            function?.acceptChildren(object : RsVisitor() {
+                override fun visitElement(element: PsiElement?) {
+                    element?.acceptChildren(this)
+                }
+
+                override fun visitFunction(o: RsFunction) {}
+                override fun visitLambdaExpr(o: RsLambdaExpr) {}
+                override fun visitRetExpr(o: RsRetExpr) = addOccurrence(o)
+
+                override fun visitTryExpr(o: RsTryExpr) {
+                    (o.expr as? RsMethodCallExpr)?.acceptChildren(this)
+                    addOccurrence(o.q)
```

Comment:
Maybe `addOccurrence(o)` would be better here. It's ok, because we don't trigger highlighting when we step onto this expression, just collect spans, so it won't overlap with the "find symbol occurrences" feature.

I think the ideal solution would be to highlight only the last function/method call in the path (to which the `?` really belongs), but the platform API doesn't allow arbitrary spans in this highlight handler (and highlighting two elements for one logical span would be incorrect), so I personally don't have strong preference here.

The current implementation also looks good to me, but feel free to return the previous logic in this place.